### DISCUSSION TITLE
Interaction between class-scoped TypeAlias and Self

### DISCUSSION THREAD

**Discussion by Gollum999** [python/typing#1526] [Q&A] [answered]
Does anyone have any thoughts on whether or not this interaction between `TypeAlias` and `Self` should be allowed?

```
from typing import Self, TypeAlias

class Foo:
    T: TypeAlias = int
    
    def bar(self, val: Self.T) -> Self.T:
        return val + 2
        
    def baz(self, val: Foo.T) -> Foo.T:
        return val * 2


Foo().bar(4)
Foo().baz(5)
```

With Python 3.11 and mypy 1.7.1, this currently results in:
```
main.py:6: error: Name "Self.T" is not defined  [name-defined]
Found 1 error in 1 file (checked 1 source file)
```

It feels to me like if `baz` works, then `bar` should also work, but I'm not sure what kind of implications that would have.  The PEPs for [`TypeAlias`](https://peps.python.org/pep-0613/) and [`Self`](https://peps.python.org/pep-0673/) don't really seem to describe how this should work.

**Comment by Gobot1234** [comment, marked as answer]
We did actually specify this in PEP 673
> We reject type aliases containing Self. Supporting Self outside class definitions can require a lot of special-handling in type checkers. Given that it also goes against the rest of the PEP to use Self outside a class definition, we believe the added convenience of aliases is not worth it:
> ```py
> TupleSelf = tuple[Self, Self]  # Rejected
> 
> class Alias:
>     def return_tuple(self) -> TupleSelf:  # Rejected
>         return (self, self)
> ```
Someone even brought it up another repo after it had been submitted (which I can't find anymore). We just decided it wasn't worth the effort. I wouldn't be particularly opposed as long as it had to be bound by a class but the whole idea of accessing members on `Self` does worry me a little because it no longer means that it can be desugared from the original bound TypeVar approach that (every??) type checker(s) use(s) to implement them.

**Reply by Gollum999** [reply]
Ahh, thanks for the context.  I do remember seeing that in the PEP, but I wasn't sure if it applied in this case since this usage of `Self` is still contained in the class definition.  But I suppose "We reject type aliases containing `Self`" is pretty explicit.
